packages/core/lib/events/DomainEventEmitter.ts

```diff
@@ -151,27 +150,69 @@ export class DomainEventEmitter<SupportedEvents extends CommonEventDefinition[]>
   public onMany<EventTypeName extends EventTypeNames<SupportedEvents[number]>>(
     eventTypeNames: EventTypeName[],
     handler: SingleEventHandler<SupportedEvents, EventTypeName>,
+    bgHandler = false,
   ) {
     for (const eventTypeName of eventTypeNames) {
-      this.on(eventTypeName, handler)
+      this.on(eventTypeName, handler, bgHandler)
     }
   }
 
   /**
    * Register handler for all events supported by the emitter
    */
-  public onAny(handler: AnyEventHandler<SupportedEvents>) {
-    this.anyHandlers.push(handler)
+  public onAny(handler: AnyEventHandler<SupportedEvents>, bgHandler = false) {
+    if (bgHandler) this.anyHandlers.background.push(handler)
+    else this.anyHandlers.foreground.push(handler)
   }
 
-  private addOnHandler(
-    eventTypeName: EventTypeNames<SupportedEvents[number]>,
-    handler: EventHandler,
-  ) {
-    if (!this.eventHandlerMap[eventTypeName]) {
-      this.eventHandlerMap[eventTypeName] = []
+  private async handleEvent<SupportedEvent extends SupportedEvents[number]>(
+    event: CommonEventDefinitionPublisherSchemaType<SupportedEvent>,
+  ): Promise<void> {
+    const eventHandlers = this.eventHandlerMap[event.type] ?? {
+      foreground: [],
+      background: [],
+    }
+
+    const fgHandlers = [...eventHandlers.foreground, ...this.anyHandlers.foreground]
+    for (const handler of fgHandlers) {
+      await handler.handleEvent(event)
     }
 
-    this.eventHandlerMap[eventTypeName].push(handler)
+    const bgHandlers = [...eventHandlers.background, ...this.anyHandlers.background]
+    for (const handler of bgHandlers) {
+      const transactionId = randomUUID()
+      // not sure if we should use startWithGroup or start, using group to group all handlers for the same event type
+      // should it be eventId + eventType or just eventType?
+      this.transactionObservabilityManager?.startWithGroup(
```

Comment:
we don't need the same thing for non-background handlers?

Reply:
Do we? it was not there before but I think it would be beneficial so will add it

Reply:
Done 🙌, thanks Igor for raising it 🙇